1/Transaction.js

```diff
@@ -68,9 +78,13 @@ class Transaction {
                     const content = Buffer.from(this._content);
 
                     sigIndexCounter++;
-                    return nacl.detached(fctUtil.sha512(Buffer.concat([index, timestamp, chainId, content])), key.secretKey);
+                    if ( key.secretKey !== undefined ) {
+                        return nacl.detached(fctUtil.sha512(Buffer.concat([index, timestamp, chainId, content])), key.secretKey);
+                    }
+                    valid = false;
+                    return undefined
                 });
-                for (let i = 0; i < this._rcds.length; i++) {
+                for (let i = 0; valid && i < this._rcds.length; i++) {
```

Comment:
if the construction of the Transaction is invalid, shouldn't we throw rather than silently fail?

Reply:
ok.  added exception.

Reply:
updated the algorithm.  there is a check for validity and no exception is required.  If no valid signature can be generated, then the transaction will require a second pass to add signatures.  It should be clearer now.